70_Pushover.pm

```diff
@@ -235,80 +245,100 @@ sub Pushover_CGI() {
             my $rDev      = "cbDev_" . $revReadings{$receipt};
 
             return ( "text/plain; charset=utf-8",
                 "NOK " . $receipt . ": invalid argument 'acknowledged'" )
-              if ( !defined( $webArgs->{acknowledged} )
-                || $webArgs->{acknowledged} ne "1" );
+              if ( defined( $webArgs->{acknowledged} )
+                && $webArgs->{acknowledged} !~ /^[01]$/ );
 
             return ( "text/plain; charset=utf-8",
                 "NOK " . $receipt . ": invalid argument 'acknowledged_by'" )
-              if ( !defined( $webArgs->{acknowledged_by} )
-                || $webArgs->{acknowledged_by} ne $hash->{USER_KEY} );
+              if ( defined( $webArgs->{acknowledged_by} )
+                && $webArgs->{acknowledged_by} !~ /^[A-Za-z0-9]{30}$/ );
+
+            return ( "text/plain; charset=utf-8",
+                "NOK " . $receipt . ": invalid argument 'acknowledged_by_device'" )
+              if ( defined( $webArgs->{acknowledged_by_device} )
+                && $webArgs->{acknowledged_by_device} !~ /^[A-Za-z0-9_-]{1,25}$/ );
 
-            if ( ReadingsVal( $name, $rAck, "1" ) eq "0"
-                && $revReadings{$receipt} > int( time() ) )
+            return ( "text/plain; charset=utf-8",
+                "NOK " . $receipt . ": invalid argument 'acknowledged_at'" )
+              if ( defined( $webArgs->{acknowledged_at} )
+                && $webArgs->{acknowledged_at} !~ /^\d{1,10}$/ );
+
+            if ( $revReadings{$receipt} > int( time() ) )
             {
                 delete $hash->{READINGS}{$rCancelId}
                   if ( defined( $hash->{READINGS}{$rCancelId} ) );
 
                 readingsBeginUpdate($hash);
 
-                readingsBulkUpdate( $hash, $rAck, "1" );
-                readingsBulkUpdate( $hash, $rAckBy,
-                    $webArgs->{acknowledged_by} );
+                if ( defined( $webArgs->{acknowledged} ) ) {
+                    readingsBulkUpdate( $hash, $rAck, $webArgs->{acknowledged} );
 
-                if ( defined( $webArgs->{acknowledged_at} )
-                    && $webArgs->{acknowledged_at} ne "" )
-                {
-                    readingsBulkUpdate( $hash, $rAckAt,
-                        $webArgs->{acknowledged_at} );
+                    if ( defined( $webArgs->{acknowledged_by} ) ) {
+                        readingsBulkUpdate( $hash, $rAckBy, $webArgs->{acknowledged_by} );
+                    }
+                    if ( defined( $webArgs->{acknowledged_by_device} ) ) {
+                        readingsBulkUpdate( $hash, $rAckByDev, $webArgs->{acknowledged_by_device} );
+                    }
+
+                    if ( defined( $webArgs->{acknowledged_at} ) ) {
+                        readingsBulkUpdate( $hash, $rAckAt, $webArgs->{acknowledged_at} );
+                    }
+                    else {
+                        readingsBulkUpdate( $hash, $rAckAt, int( time() ) );
+                    }
```

Comment:
Only set the the various acknowledge readings if the message has actually been acknowledged.